pennylane/qnn/iqp.py

```diff
@@ -0,0 +1,387 @@
+        except TypeError:
+            shape = (samples_len,)
+
+        return 2 - jnp.repeat(ops_sum, samples_len).reshape(shape) % 2 - 2 * (samples_sum % 2)
+
+    return 1
+
+
+# pylint: disable=too-many-arguments
+def _op_expval_batch(
+    gates: list,
+    params: list,
+    n_qubits: int,
+    ops: list,
+    n_samples: int,
+    key: list,
+    spin_sym: bool = False,
+    sparse: bool = False,
+    indep_estimates: bool = False,
+) -> list:
+
+    if indep_estimates:
+        return _op_expval_indep(
+            gates=gates,
+            n_qubits=n_qubits,
+            params=params,
+            ops=ops,
+            n_samples=n_samples,
+            key=key,
+            sparse=sparse,
+            spin_sym=spin_sym,
+        )
+
+    samples = jax.random.randint(key, (n_samples, n_qubits), 0, 2)
+
+    generators = _generators(gates, n_qubits)
+    effective_params = _effective_params(gates, params)
+
+    generators_sp = None
+    if sparse:
+        generators_sp = _generators_sp(gates, n_qubits)
+
+    if sparse or isinstance(ops, csr_matrix):
+        samples, generators_sp, ops = _to_csr(generators, ops, samples, generators_sp)
+
+        samples_gates, samples_sum, samples_len = _sparse_samples(generators_sp, samples, spin_sym)
+        ops_gen, ops_sum = _sparse_ops(generators_sp, ops, spin_sym)
+
+    else:
+        samples_gates, samples_sum, samples_len = _dense_samples(generators, samples, spin_sym)
+        ops_gen, ops_sum = _dense_ops(generators, ops, spin_sym)
+
+    ini_spin_sym = _ini_spin_sym(ops_sum, samples_sum, samples_len, spin_sym)
+
+    par_ops_gates = 2 * effective_params * ops_gen
+    expvals = ini_spin_sym * jnp.cos(par_ops_gates @ samples_gates.T)
+
+    return expvals
+
+
+# pylint: disable=too-many-arguments
+def iqp_expval(
+    ops: list,
+    weights: list[float],
+    pattern: list[list[list[int]]],
+    num_wires: int,
+    n_samples: int,
+    key: list,
+    spin_sym: bool = False,
+    sparse: bool = False,
+    indep_estimates: bool = False,
+    max_batch_ops: int = None,
+    max_batch_samples: int = None,
+) -> list:
+    r"""Estimate the expectation values of a batch of Pauli-Z type operators. A set of l operators must be specified
+    by an array of shape (l,n_qubits), where each row is a binary vector that specifies on which qubit a Pauli Z
+    operator acts.
+    The expectation values are estimated using a randomized method whose precision in controlled by n_samples,
+    with larger values giving higher precision. Estimates are unbiased, however may be correlated. To request
+    uncorrelated estimate, use indep_estimates=True at the cost of larger runtime.
+    For large batches of operators or large values of n_samples, memory can be controlled by setting max_batch_ops
+    and/or max_batch_samples to a fixed value.
+
+    Args:
+        ops (jnp.ndarray): Array specifying the operator/s for which to estimate the expectation values.
```

Comment:
Does the `ops` type has to be `jnp.ndarray`?
I used a numpy array in an example and it seems to also work